sphinx/transforms/post_transforms/__init__.py

```diff
@@ -246,18 +246,23 @@ def has_visitor(translator: type[nodes.NodeVisitor], node: type[Element]) -> boo
             # subclass of SphinxTranslator supports desc_sig_element nodes automatically.
             return
 
-        # for the leaf elements (desc_sig_element), the translator should support _all_
-        if not all(has_visitor(translator, node) for node in addnodes.SIG_ELEMENTS):
+        # for the leaf elements (desc_sig_element), the translator should support _all_,
+        # unless there exists a generic visit_desc_sig_element default visitor
+        if (not all(has_visitor(translator, node) for node in addnodes.SIG_ELEMENTS)
```

Comment:
equivalent logic, but this version would short-circuit faster:

```suggestion
        if (any(not has_visitor(translator, node) for node in addnodes.SIG_ELEMENTS)
```

Reply:
Could you enlighten me why this is the case? For me both short-circuit at the same item yield False. However, the bytecode of my version seems to be more efficient:

```sh
$ read -r -d '' code << EOM
import dis

def pred(x): ...

def foo(X):
    if not all(pred(x) for x in X):
        return 1
    return 2

def bar(X):
    if any(not pred(x) for x in X):
        return 1
    return 2

print('dis.dis(foo)')
dis.dis(foo)
print()
print('dis.dis(bar)')
dis.dis(bar)
EOM

$ python3.9 -c "$code" 
dis.dis(foo)
  6           0 LOAD_GLOBAL              0 (all)
              2 LOAD_CONST               1 (<code object <genexpr> at 0x7f316ca149d0, file "<string>", line 6>)
              4 LOAD_CONST               2 ('foo.<locals>.<genexpr>')
              6 MAKE_FUNCTION            0
              8 LOAD_FAST                0 (X)
             10 GET_ITER
             12 CALL_FUNCTION            1
             14 CALL_FUNCTION            1
             16 POP_JUMP_IF_TRUE        22

  7          18 LOAD_CONST               3 (1)
             20 RETURN_VALUE

  8     >>   22 LOAD_CONST               4 (2)
             24 RETURN_VALUE

Disassembly of <code object <genexpr> at 0x7f316ca149d0, file "<string>", line 6>:
  6           0 LOAD_FAST                0 (.0)
        >>    2 FOR_ITER                14 (to 18)
              4 STORE_FAST               1 (x)
              6 LOAD_GLOBAL              0 (pred)
              8 LOAD_FAST                1 (x)
             10 CALL_FUNCTION            1
             12 YIELD_VALUE
             14 POP_TOP
             16 JUMP_ABSOLUTE            2
        >>   18 LOAD_CONST               0 (None)
             20 RETURN_VALUE

dis.dis(bar)
 11           0 LOAD_GLOBAL              0 (any)
              2 LOAD_CONST               1 (<code object <genexpr> at 0x7f316c77c030, file "<string>", line 11>)
              4 LOAD_CONST               2 ('bar.<locals>.<genexpr>')
              6 MAKE_FUNCTION            0
              8 LOAD_FAST                0 (X)
             10 GET_ITER
             12 CALL_FUNCTION            1
             14 CALL_FUNCTION            1
             16 POP_JUMP_IF_FALSE       22

 12          18 LOAD_CONST               3 (1)
             20 RETURN_VALUE

 13     >>   22 LOAD_CONST               4 (2)
             24 RETURN_VALUE

Disassembly of <code object <genexpr> at 0x7f316c77c030, file "<string>", line 11>:
 11           0 LOAD_FAST                0 (.0)
        >>    2 FOR_ITER                16 (to 20)
              4 STORE_FAST               1 (x)
              6 LOAD_GLOBAL              0 (pred)
              8 LOAD_FAST                1 (x)
             10 CALL_FUNCTION            1
             12 UNARY_NOT
             14 YIELD_VALUE
             16 POP_TOP
             18 JUMP_ABSOLUTE            2
        >>   20 LOAD_CONST               0 (None)
             22 RETURN_VALUE
```

In particular, here's what changes:

```python
Disassembly of <code object <genexpr> at 0x7f9df4519a80, file "<string>", line 6>:
  6           0 LOAD_FAST                0 (.0)
        >>    2 FOR_ITER                14 (to 18)
              4 STORE_FAST               1 (x)
              6 LOAD_GLOBAL              0 (pred)
              8 LOAD_FAST                1 (x)
             10 CALL_FUNCTION            1
             12 YIELD_VALUE
             14 POP_TOP
             16 JUMP_ABSOLUTE            2
        >>   18 LOAD_CONST               0 (None)
             20 RETURN_VALUE
```

```python
Disassembly of <code object <genexpr> at 0x7f9df4519f50, file "<string>", line 11>:
 11           0 LOAD_FAST                0 (.0)
        >>    2 FOR_ITER                16 (to 20)
              4 STORE_FAST               1 (x)
              6 LOAD_GLOBAL              0 (pred)
              8 LOAD_FAST                1 (x)
             10 CALL_FUNCTION            1
             12 UNARY_NOT                <- due to 'any(not ...)'
             14 YIELD_VALUE
             16 POP_TOP
             18 JUMP_ABSOLUTE            2
        >>   20 LOAD_CONST               0 (None)
             22 RETURN_VALUE
```

There is one `UNARY_NOT` instruction due to `any(not ...)` which is not optimized. Since `all` and `any` shortcircuit similarly, I would say that my version is faster (essentially, the "not pred(...)" seems to not be optimized in the bytecode).

Reply:
You're quite right -- I was only thinking about the negative case, not the positive. Sorry!